clients/drcachesim/scheduler/scheduler_impl.h

```diff
@@ -0,0 +1,985 @@
+        output_ordinal_t prev_output = sched_type_t::INVALID_OUTPUT_ORDINAL;
+        // The current output where we're actively running.
+        output_ordinal_t cur_output = sched_type_t::INVALID_OUTPUT_ORDINAL;
+        uintptr_t next_timestamp = 0;
+        uint64_t instrs_in_quantum = 0;
+        int instrs_pre_read = 0;
+        // This is a per-workload value, stored in each input for convenience.
+        uint64_t base_timestamp = 0;
+        // This equals 'options_.deps == DEPENDENCY_TIMESTAMPS', stored here for
+        // access in InputTimestampComparator which is static and has no access
+        // to the schedule_t.  (An alternative would be to try to get a lambda
+        // with schedule_t "this" access for the comparator to compile: it is not
+        // simple to do so, however.)
+        bool order_by_timestamp = false;
+        // Global queue counter used to provide FIFO for same-priority inputs.
+        // This value is only valid when this input is in a queue; it is set upon
+        // being added to a queue.
+        uint64_t queue_counter = 0;
+        // Used to switch on the instruction *after* a long-latency syscall.
+        bool processing_syscall = false;
+        bool processing_maybe_blocking_syscall = false;
+        uint64_t pre_syscall_timestamp = 0;
+        // Use for special kernel features where one thread specifies a target
+        // thread to replace it.
+        input_ordinal_t switch_to_input = sched_type_t::INVALID_INPUT_ORDINAL;
+        uint64_t syscall_timeout_arg = 0;
+        // Used to switch before we've read the next instruction.
+        bool switching_pre_instruction = false;
+        // Used for time-based quanta.  The units are simulation time.
+        uint64_t prev_time_in_quantum = 0;
+        uint64_t time_spent_in_quantum = 0;
+        // These fields model waiting at a blocking syscall.
+        // The units are in simuilation time.
+        uint64_t blocked_time = 0;
+        uint64_t blocked_start_time = 0;
+        // An input can be "unscheduled" and not on the ready_priority_ run queue at all
+        // with an infinite timeout until directly targeted.  Such inputs are stored
+        // in the unscheduled_priority_ queue.
+        // This field is also set to true for inputs that are "unscheduled" but with
+        // a timeout, even though that is implemented by storing them in ready_priority_
+        // (because that is our mechanism for measuring timeouts).
+        bool unscheduled = false;
+        // Causes the next unscheduled entry to abort.
+        bool skip_next_unscheduled = false;
+        uint64_t last_run_time = 0;
+    };
+
+    // Format for recording a schedule to disk.  A separate sequence of these records
+    // is stored per output stream; each output stream's sequence is in one component
+    // (subfile) of an archive file.
+    // All fields are little-endian.
+    START_PACKED_STRUCTURE
+    struct schedule_record_t {
```

Comment:
Should this be part of the private implementation or the public header?

Reply:
I think it should be private.  We should provide tools if we want users to look inside schedule files; the files should be treated as opaque.